File: modules/08-api-boundaries/swift/Sources/Account/Account.swift
// Module 08 in Swift: the richest access-control ladder of the five --
// `private` (enclosing declaration), `fileprivate` (the file), `internal`
// (the module, the default), `public` (other modules), and `open` (other
// modules may subclass/override). Account is a public struct whose stored
// property is `private`, so consumers use the methods but cannot touch or
// construct balance directly.

public struct Overdraft: Error {}

public struct Account {
  private var balance: Int64

  public init(_ initial: Int64) throws {
    if initial < 0 { throw Overdraft() }
    balance = initial
  }

  public mutating func deposit(_ amount: Int64) { balance += amount }

  public mutating func withdraw(_ amount: Int64) throws {
    if amount > balance { throw Overdraft() }
    balance -= amount
  }

  public func value() -> Int64 { balance }
}

File: modules/08-api-boundaries/ocaml/lib/account.ml
(* The implementation. The record type is concrete HERE, but because account.mli
   exposes `type t` abstractly, none of this representation escapes the module.
   Values are immutable, so deposit/withdraw return a new t. *)

type t = { balance : int }

exception Overdraft

let open_ initial = if initial < 0 then raise Overdraft else { balance = initial }
let deposit a amount = { balance = a.balance + amount }

let withdraw a amount =
  if amount > a.balance then raise Overdraft else { balance = a.balance - amount }

let balance a = a.balance
